OpenRA.Mods.Common/Traits/BotModules/SendUnitToAttackBotModule.cs

```diff
@@ -0,0 +1,183 @@
+	{
+		[Desc("Actors used for attack, and their base desire provided for attack desire.",
+			"When desire reach 100, AI will send them to attack.")]
+		public readonly Dictionary<string, int> ActorTypesAndAttackDesire = default;
+
+		[Desc("Target types that can be targeted.")]
+		public readonly BitSet<TargetableType> ValidTargets = new("Structure");
+
+		[Desc("Target types that can't be targeted.")]
+		public readonly BitSet<TargetableType> InvalidTargets;
+
+		[Desc("Should attack the furthest or closest target. Possible values are Closest, Furthest, Random")]
+		public readonly AttackDistance AttackDistance = AttackDistance.Closest;
+
+		[Desc("Attack order name.")]
+		public readonly string AttackOrderName = "Attack";
+
+		[Desc("Find target and try attack target in this interval.")]
+		public readonly int ScanTick = 463;
+
+		[Desc("The total attack desire increase this amount per scan",
+		"Note: when there is no attack unit, total attack desire will return 0.")]
+		public readonly int AttackDesireIncreasedPerScan = 10;
+
+		public override object Create(ActorInitializer init) { return new SendUnitToAttackBotModule(init.Self, this); }
+	}
+
+	public class SendUnitToAttackBotModule : ConditionalTrait<SendUnitToAttackBotModuleInfo>, IBotTick
+	{
+		readonly World world;
+		readonly Player player;
+		readonly Predicate<Actor> unitCannotBeOrdered;
+		readonly Predicate<Actor> unitCannotBeOrderedOrIsBusy;
+		readonly Predicate<Actor> isInvalidActor;
+		int minAssignRoleDelayTicks;
+		Player targetPlayer;
+		int desireIncreased;
+
+		public SendUnitToAttackBotModule(Actor self, SendUnitToAttackBotModuleInfo info)
+		: base(info)
+		{
+			world = self.World;
+			player = self.Owner;
+			isInvalidActor = a => a == null || a.IsDead || !a.IsInWorld || a.Owner != targetPlayer;
+			unitCannotBeOrdered = a => a == null || a.IsDead || !a.IsInWorld || a.Owner != player;
+			unitCannotBeOrderedOrIsBusy = a => unitCannotBeOrdered(a) || (!a.IsIdle && a.CurrentActivity is not FlyIdle);
+			desireIncreased = 0;
+		}
+
+		protected override void TraitEnabled(Actor self)
+		{
+			// Avoid all AIs reevaluating assignments on the same tick, randomize their initial evaluation delay.
+			minAssignRoleDelayTicks = world.LocalRandom.Next(0, Info.ScanTick);
+		}
+
+		void IBotTick.BotTick(IBot bot)
+		{
+			if (--minAssignRoleDelayTicks <= 0)
+			{
+				minAssignRoleDelayTicks = Info.ScanTick;
+
+				var attackdesire = 0;
+				var actors = world.ActorsWithTrait<IPositionable>().Select(at => at.Actor).Where(a =>
+				{
+					if (Info.ActorTypesAndAttackDesire.ContainsKey(a.Info.Name) && !unitCannotBeOrderedOrIsBusy(a))
+					{
+						attackdesire += Info.ActorTypesAndAttackDesire[a.Info.Name];
+						return true;
+					}
+
+					return false;
+				}).ToList();
+
+				if (actors.Count == 0)
+					desireIncreased = 0;
+				else
+					desireIncreased += Info.AttackDesireIncreasedPerScan;
+
+				if (desireIncreased + attackdesire < 100)
```

Comment:
The entire "desire" system is a bit weird. Why not just send the unit as soon as it is available?

Reply:
Because in TS, we will want not only 1 subtank to attack, if possible AI can wait for more subtanks.

If you set it to 100, it becomes send the unit as soon as it is available.

Reply:
That sounds like subtanks should have their own squad logic?

Reply:
Targeting the furthest structure seems to be independent of the desire system.

Reply:
Then think about following usecase: we have bomb truck but generate smaller explosion, which means they will only be threatening when in mass. In this case, the desire system will also in need.

Reply:
For example, the bomber pod in OpenHV.

Reply:
So the desire system is a unit number controller: if 1 subtank is threatening enough, we can ignore the desire system, and it is the same for mad tank and demo truck. Otherwise, the desire system will be in needed to help AI gather enough units.

Reply:
In conclusion, 

This module is for "units that should not be sent with a regular squad".

Desire system is to control the number of those units.

Targeting system is to control the target of those units.

Sometimes the unit is dangerous enough alone, which looks like desire system is no longer needed, but actually for better usage we need the desire system.

Reply:
desire should be per unit type, not per trait. Otherwise you are basically forced to add a trait for each unit type

Reply:
`IPosition` is just a filter to find out actor that is valid. See the same usage in UnitBuilderBotModule.